### PR TITLE
collect_unicodes() to check gid < num_glyphs with cmap 12

### LINKED ISSUE #2204: A shape fuzzer timeout case
I have apparently <s>a recent regression</s> (maybe not) of an infinite loop with this font,

[1746cad6bc3fb2b355db50a5af37c9b58d9ad376.zip](https://github.com/harfbuzz/harfbuzz/files/4263528/1746cad6bc3fb2b355db50a5af37c9b58d9ad376.zip)

`clang test/fuzzing/hb-shape-fuzzer.cc test/fuzzing/main.cc -Isrc src/harfbuzz.cc -DHB_NO_MT -lm -fno-exceptions -g -Og && ./a.out 1746cad6bc3fb2b355db50a5af37c9b58d9ad376`

```
Program received signal SIGINT, Interrupt.
0x00005555555b8ee5 in hb_set_t::get_major (this=0x555555710f70, 
    g=<optimized out>) at src/hb-set.hh:876
876	  unsigned int get_major (hb_codepoint_t g) const { return g / page_t::PAGE_BITS; }
(gdb) bt
#0  0x00005555555b8ee5 in hb_set_t::get_major (this=0x555555710f70, 
    g=<optimized out>) at src/hb-set.hh:876
#1  0x00005555555b8f12 in hb_set_t::page_for_insert (this=0x555555710f70, 
    g=361016) at src/hb-set.hh:843
#2  0x0000555555594bd5 in hb_set_t::add (this=0x555555710f70, g=361016)
    at src/hb-set.hh:289
#3  0x00005555555a16f2 in OT::CmapSubtableLongSegmented<OT::CmapSubtableFormat12>::collect_unicodes (this=<optimized out>, out=0x555555710f70)
    at src/hb-ot-cmap-table.hh:554
#4  0x00005555555565d7 in test_font (font=0x55555570fb00, cp=1744896000)
    at test/fuzzing/../api/test-ot-face.c:48
#5  0x0000555555556532 in LLVMFuzzerTestOneInput (data=<optimized out>, 
    size=23293) at test/fuzzing/hb-shape-fuzzer.cc:53
#6  0x0000555555556a2d in main (argc=2, argv=0x7fffffffdc58)
    at test/fuzzing/main.cc:15
```
Pretty sure this is a recent regression, can you @garretrieger @blueshade7 have a quick look at it? Thanks

### PR DESCRIPTION
**Pull request by blueshade7**
fixes #2204 "Infinite loop on shape fuzzer"

### CODE REVIEW

File: src/hb-ot-cmap-table.hh [hunk 1]
```diff
@@ -576,9 +576,13 @@ struct CmapSubtableLongSegmented
 struct CmapSubtableFormat12 : CmapSubtableLongSegmented<CmapSubtableFormat12>
 {
   static hb_codepoint_t group_get_glyph (const CmapSubtableLongGroup &group,
-					 hb_codepoint_t u)
-  { return likely (group.startCharCode <= group.endCharCode) ?
-	   group.glyphID + (u - group.startCharCode) : 0; }
+					 hb_codepoint_t u, unsigned int num_glyphs = UINT_MAX)
+  { hb_codepoint_t g = group.glyphID;
```
Comment (ebraminio) [inline src/hb-ot-cmap-table.hh]: Feel free to use new line here, we use curly braces in the same for one liner functions.

File: src/hb-ot-cmap-table.hh [hunk 2]
```diff
@@ -652,8 +656,8 @@ struct CmapSubtableFormat12 : CmapSubtableLongSegmented<CmapSubtableFormat12>
 struct CmapSubtableFormat13 : CmapSubtableLongSegmented<CmapSubtableFormat13>
 {
   static hb_codepoint_t group_get_glyph (const CmapSubtableLongGroup &group,
-					 hb_codepoint_t u HB_UNUSED)
-  { return group.glyphID; }
+					 hb_codepoint_t u HB_UNUSED, unsigned int num_glyphs = UINT_MAX)
+  { return likely (group.glyphID < num_glyphs)? group.glyphID: 0; }
```
Comment (ebraminio) [inline src/hb-ot-cmap-table.hh]: Please put space before `?` and `:` :)

**Comment by ebraminio** [comment]
> fixes issue #2204 "Infinite loop on shape fuzzer"

Edited and used `fixes #2204`, if that (or `closes #2204`) is put on commit's summary or title it will close the file when a PR is merged.

**Comment by blueshade7** [comment]
Edited reflecting comments. Thanks.

**Comment by ebraminio** [comment]
Please put the font that shows the issue on test/fuzzing/fonts :)

**Comment by blueshade7** [comment]
It is still iterating over out of range glyphs, taking a few seconds to complete.
A possible further improvement may be to clamp the range first to not call `group_get_glyph()` for those in the first place.

**Comment by ebraminio** [comment]
Feel free to upload a patch for that also :)

**Comment by ebraminio** [comment]
Causing bots to fail, reverted https://github.com/harfbuzz/harfbuzz/commit/e90213868b121af72d19b6022a671ebe72ff55aa :/ let's continue on #2204

**Pull request merged**